File: leetcode/Linked List/#160-Intersection of Two Linked Lists/main.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        countA, countB = 0, 0
        temp1, temp2 = headA, headB

        if temp1 == None or temp2 == None: return None
        while temp1 != None or temp2 != None:
            if temp1 != None:
                temp1 = temp1.next
                countA += 1
            if temp2 != None:
                temp2 = temp2.next
                countB += 1

        temp1, temp2 = headA, headB

        if countA > countB:
            for s in range(countA - countB): temp1 = temp1.next
            for a in range(countB):
                if temp1 == temp2: return temp1
                temp1, temp2 = temp1.next, temp2.next
        else:
            for s in range(countB - countA): temp2 = temp2.next
            for a in range(countA):
                if temp1 == temp2: return temp2
                temp1, temp2 = temp1.next, temp2.next

        return None

File: leetcode/Linked List/#160-Intersection of Two Linked Lists/main.cpp

 // * Definition for singly-linked list.
 
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (headA == nullptr || headB == nullptr) {
            return nullptr;
        }

        ListNode *ptrA = headA;
        ListNode *ptrB = headB;

        int LA = 1;
        int LB = 1;

        while (ptrA->next != nullptr) {
            ptrA = ptrA->next;
            LA++;
        }

        while (ptrB->next != nullptr) {
            ptrB = ptrB->next;
            LB++;
        }

        if (ptrA != ptrB) {
            return nullptr;
        }

        ptrA = headA;
        ptrB = headB;
        int dist = 0;

        if (LA > LB) {
            dist = LA - LB;
            while (dist > 0) {
                ptrA = ptrA->next;
                dist--;
            }

            while ((ptrA != nullptr) && (ptrB != nullptr)) {
                if (ptrA == ptrB) {
                    return ptrA;
                }

                ptrA = ptrA->next;
                ptrB = ptrB->next;
            }
        }

        else {
            dist = LB - LA;
            while (dist > 0) {
                ptrB = ptrB->next;
                dist--;
            }

            while ((ptrA != nullptr) && (ptrB != nullptr)) {
                if (ptrA == ptrB) {
                    return ptrA;
                }

                ptrA = ptrA->next;
                ptrB = ptrB->next;
            }
        }



    }
};
